### ISSUE TITLE
Loss is negative when use SparsemaxLoss to do 1 of K prediction

### ISSUE THREAD

**Issue by bigtreeshen** [gokceneraslan/SparseMax.torch#2]
Hi gokceneraslan,

Thanks for your code of SparseMax! I got a problem when I transfer these code to resnet (I use facebook resnet code https://github.com/facebook/fb.resnet.torch) and my task is just like ImageNet classification task which is a 1 of K Classification, but I got negative Loss like follows:

 Epoch: [1][2102/22574]    Time 0.491  Data 0.000  Err -0.4587  top1 100.000  top5 100.000

I don't know whether SparseMaxLoss is suitable for 1 of K Classification task since in the original paper SparseMaxLoss is used for Mutillabel Classification. Or do I need to modify your code to make it suitable for 1 of K Classification task?

**Comment by allanzelener** [comment]
SparseMaxLoss should be suitable for 1 of K classification. Are you using SparseMaxLoss with ClassNLLCriterion? The SparseMaxLoss module is only a scoring function like LogSoftMax, it does not define a criterion to train a model.

Also to be honest, I haven't thoroughly tested the SparseMaxLoss implementation on a benchmark ML task. It may be a good idea to try it on MNIST classification first before training a larger model.

**Comment by bigtreeshen** [comment]
@allanzelener  Thanks for your advice! And yes, I am using SML with ClassNLLCCriterion just like the code in SparseMaxCriterion.lua. I also use TestSparseMax.lua to have a simple test, here is my result:

`require 'models/SparseMax'`
`require 'models/SparseMaxLoss'`
`sml = nn.SparseMaxLoss()`
`nll = nn.ClassNLLCriterion()`

`x = torch.randn(5)`
`x`
`0.5366`
`0.2463`
`0.6123`
`2.2302`
`0.4678`

`x_sml = sml:forward(x)`
`x_sml`
`1.3333`
`1.0430`
`1.4090`
`3.0269`
`1.2645`

`y = 2`

`loss = nll:forward(x_sml, y)`
`loss`
`-1.0429894251011`
here I got a negative loss result, did I do anything wrong?

**Comment by bigtreeshen** [comment]
Dear @allanzelener  and @gokceneraslan ,

I think there is a little part need to be modified in SparseMaxLoss.lua, 

In SparseMaxLoss.lua, from line52 - 54， the original code is like this:

`local zs_squared = torch.cmul(zs_sparse, zs_sparse)`
`local taus_squared = torch.cmul(self.taus, self.taus):expandAs(zs_sparse)`
`local regTerm = (torch.sum(zs_squared - taus_squared, dim) * 0.5 + 0.5):expandAs(input)`

I think it should be like this 
`local zs_squared = torch.cmul(zs_sparse, zs_sparse)`
`local taus_squared = torch.cmul(self.taus, self.taus):expandAs(zs_sparse)`
`taus_squared = torch.cmul(is_gt, taus_squared)`
`local regTerm = (torch.sum(zs_squared - taus_squared, dim) * 0.5 + 0.5):expandAs(input)`

Correct me if I am wrong. Thank you very much!

**Comment by allanzelener** [comment]
Yes, you're right. `regTerm` should only sum over the support of the input so `taus_squared` should be masked after being expanded.

I added your fix in this commit: dd781e84f7d774ef36a5a94a369917da39c764e8

Please let me know if there's any other problems. I'll leave this issue open for now until `SparseMaxLoss` is more thoroughly tested on a benchmark task.

Edit: Okay, I've tried `SparseMaxLoss` as a drop-in replacement for `LogSoftMax` for the [Torch MNIST demo](https://github.com/torch/demos/blob/master/train-a-digit-classifier/train-on-mnist.lua) and it seems to train reasonably well. Thanks for helping find the bug!

**Comment by bigtreeshen** [comment]
@allanzelener  Sure, I also run the code on my own dataset, it seems good. Thanks for your work!
